sim/build.sbt

```diff
@@ -4,6 +4,30 @@ import Tests._
 val apiDirectory = settingKey[String]("The site directory into which the published scaladoc should placed.")
 apiDirectory := "latest"
 
+def filenames(tempDir: File, fs: Seq[File]): Seq[String] =
+for(f <- fs) yield {
+  sbtassembly.AssemblyUtils.sourceOfFileForMerge(tempDir, f) match {
+    case (path, base, subDirPath, false) => subDirPath
+    case (jar, base, subJarPath, true) => jar + ":" + subJarPath
+  }
+}
+
+val not_rc_merge_strategy = new sbtassembly.MergeStrategy {
+  val name = "not_rocket"
+  def apply(tempDir: File, path: String, files: Seq[File]) = {
+    val filtered = files collect { f =>
+      sbtassembly.AssemblyUtils.sourceOfFileForMerge(tempDir, f) match {
+        case (jar, _, _, true) if !jar.toString.contains("rocketchip") => f
+      }
+    }
+    if (filtered.size == 1) Right(Seq(filtered.head -> path))
+    else Left("still have multiple files after removing rocketchip for same target path:" +
+      filenames(tempDir, filtered).mkString("\n", "\n", "")
+      )
+  }
+}
```

Comment:
It seems like `emulator.cc` should come from whatever isn't rocketchip because it was overridden.  I added a custom merge strategy to the firesim/sim/build.sbt in 0c55c1c.  I don't like how hacky it looks.  Perhaps @jackkoenig might have some advice on making it more concise?  This should make the CI pass.

Reply:
Thanks @jackkoenig , I get confused about whether it's better to use `filter`, `map` or `collect`.  My understanding is that `collect` is just a combination of `filter` and `map`.

Now that I have the CI passing for FireSim using the fat jars and after looking at [chipyard/generators/utilities/src/main/scala/Simulator.scala's use of `getClass.getResourceAsStream()`](https://github.com/ucb-bar/chipyard/blob/02c889b8b15b86587feae852f1165fade348bb4b/generators/utilities/src/main/scala/Simulator.scala#L68-L75) to emit `/csrc/emulator.cc`, we definitely want to make sure that `chipyard/build.sbt` also updates it's `assemblyMergeStrategy` to be a bit more conservative and also handle picking `emulator.cc` that isn't coming from rocketchip.  I'll open a dependent Chipyard PR.

Reply:
Pretty much the only case i use filter + map these days is if i can _very_ succinctly describe a function (T) => Bool. Otherwise i use collect. It really shines if you need to do any non-trivial pattern matching or want to filter on type, and tends to be more succinct. 

In general you should probably tend to using fewer higher-order functions for readability.